Add skip entries option

svartkanin opened the pull request:
This addresses #45 

I created a new menu option `skip_empty_entries` that when set will skip over empty entries in the menu. 

Skip-able entries can either be defined with empty string or as None. The new option is set to `False` as default, so if empty or None values or specified it will fail as it does now. But if the option is set to `True` it will skip the entries when iterating through the menu. This will allow for grouping menu options together visually. 

I've tested it with single selection and multiselection.

IngoMeyer441 commented:
Thank you very much for contributing! The changes look good to me. I have only some suggestions for minor improvements:

- Declaring the type of a variable without a typing comment is "only" possible for Python >= 3.6. Some people requested Python 3.5 compatibility, so I still use the old typing comments (see the *Python 3.5 compatibility check* in the GitHub actions).
- I think it is a good idea to both allow `None` and empty strings for menu separators because empty strings could also be passed from the command line interface. Since `None` is a special value which is invalid without your new flag, I think we could even always use `None` as a separator entry. This simplifies the API usage.
- The pre-commit action failed because of a broken `black` installation... I fixed this in the latest `develop`, so this check should pass when you push again.
- It would be great if you could add a sentence to the `README.md` file which describes your new flag. I use the `README.md` as a documentation replacement. :innocent:

svartkanin commented:
Thanks for the quick review and reply.
I'm happy to address them, probably towards the end of the weekend :)

svartkanin commented:
@IngoMeyer441 I've updated the PR now: 
- Added type comments to the declared variables
- Only use `None` as the separator 
- Updated the README.md with an example
- Also in the first version I forgot to add the command line argument for the new constructor parameter, that is included as well now

I had issues running the black commit hook. I kept getting the following error 
```
Traceback (most recent call last):
  File ".cache/pre-commit/repootteniip/py_env-python3/bin/black", line 8, in <module>
    sys.exit(patched_main())
  File ".cache/pre-commit/repootteniip/py_env-python3/lib/python3.10/site-packages/black/__init__.py", line 1423, in patched_main
    patch_click()
  File ".cache/pre-commit/repootteniip/py_env-python3/lib/python3.10/site-packages/black/__init__.py", line 1409, in patch_click
    from click import _unicodefun
ImportError: cannot import name '_unicodefun' from 'click' (.cache/pre-commit/repootteniip/py_env-python3/lib/python3.10/site-packages/click/__init__.py)
```

I updated the hook to the latest version of black and that resolved the issue

IngoMeyer441 commented:
Thanks for your changes and your contribution. I tested your code and works fine!
I applied some minor improvements to your PR:

- Always consider `None` as an empty menu entry (independently from `skip_empty_entries`)
- Consider empty strings as empty menu items if `skip_empty_entries` is
  set to `True`
- Do not require preview functions to accept a `None` value
- Replace `Union[None, x]` with `Optional[x]`
- Sort constructor parameters and cli switches
- Update the documentation of command line switches

I pushed a new commit to the `svartkanin-skip-empty-entries` branch in this repository. You can apply (and review) the changes with:

```bash
git fetch https://github.com/IngoMeyer441/simple-term-menu.git svartkanin-skip-empty-entries
git cherry-pick FETCH_HEAD
```

If the changes are ok for you, simply push the commit to this PR and I will merge it. Otherwise, please let me know. I am always open for discussion. :slightly_smiling_face: 

Thanks again and have a nice weekend! :beers:

svartkanin commented:
@IngoMeyer441 thanks for the updates, I've tested things and it seems to work as expected :+1:
